src/pods/generate/TemplatedResourcesGenerator.ts

```diff
@@ -0,0 +1,98 @@
+import { promises as fsPromises } from 'fs';
+import { posix } from 'path';
+import type { Representation } from '../../ldp/representation/Representation';
+import { RepresentationMetadata } from '../../ldp/representation/RepresentationMetadata';
+import type { ResourceIdentifier } from '../../ldp/representation/ResourceIdentifier';
+import { getLoggerFor } from '../../logging/LogUtil';
+import type { FileIdentifierMapper, FileIdentifierMapperFactory } from '../../storage/mapping/FileIdentifierMapper';
+import { guardedStreamFrom } from '../../util/StreamUtil';
+import type { ResourcesGenerator } from './ResourcesGenerator';
+import type { TemplateEngine } from './TemplateEngine';
+import Dict = NodeJS.Dict;
+
+const { join: joinPath } = posix;
+
+/**
+ * Generates resources by making use of a template engine.
+ * The template folder structure will be kept.
+ * Folders will be interpreted as containers and files as documents.
+ * A FileIdentifierMapper will be used to generate identifiers that correspond to the relative structure.
+ */
+export class TemplatedResourcesGenerator implements ResourcesGenerator {
+  protected readonly logger = getLoggerFor(this);
+
+  private readonly templateFolder: string;
+  private readonly factory: FileIdentifierMapperFactory;
+  private readonly engine: TemplateEngine;
+
+  /**
+   * A mapper is needed to convert the template file paths to identifiers relative to the given base identifier.
+   *
+   * @param templateFolder - Folder where the templates are located.
+   * @param factory - Factory used to generate mapper relative to the base identifier.
+   * @param engine - Template engine for generating the resources.
+   */
+  public constructor(templateFolder: string, factory: FileIdentifierMapperFactory, engine: TemplateEngine) {
+    this.templateFolder = templateFolder;
+    this.factory = factory;
+    this.engine = engine;
+  }
+
+  public async generate(location: ResourceIdentifier, options: Dict<string>):
+  Promise<Map<ResourceIdentifier, Representation>> {
+    const mapper = await this.factory.create(location.path, this.templateFolder);
+    const result = new Map<ResourceIdentifier, Representation>();
+    await this.parseFolder(this.templateFolder, mapper, options, result);
```

Comment:
Return the result instead of passing it?

Reply:
There is no easy way to join maps besides iterating over all the keys and adding them so that's why I went for this. Might still change it for an async generator as mentioned above anyway.